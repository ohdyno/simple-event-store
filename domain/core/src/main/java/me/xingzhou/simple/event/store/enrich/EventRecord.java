package me.xingzhou.simple.event.store.enrich;

import me.xingzhou.simple.event.store.Event;
import me.xingzhou.simple.event.store.RecordDetails;
import me.xingzhou.simple.event.store.storage.StoredRecord;

public record EventRecord(Event event, RecordDetails details) {
    public static EventRecord extract(StoredRecord record, Event event) {
        return new EventRecord(
                event, new RecordDetails(record.streamName(), record.version(), record.id(), record.insertedOn()));
    }
}
