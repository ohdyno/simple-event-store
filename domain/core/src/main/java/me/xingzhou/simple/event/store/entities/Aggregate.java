package me.xingzhou.simple.event.store.entities;

import me.xingzhou.simple.event.store.ids.StreamName;
import me.xingzhou.simple.event.store.ids.Version;

public interface Aggregate extends EventSourceEntity {
    void setVersion(Version version);

    StreamName streamName();

    Version version();
}
