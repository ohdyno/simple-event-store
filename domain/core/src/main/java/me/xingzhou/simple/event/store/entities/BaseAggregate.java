package me.xingzhou.simple.event.store.entities;

import me.xingzhou.simple.event.store.ids.Version;
import me.xingzhou.simple.event.store.storage.EventStorage;

public abstract class BaseAggregate implements Aggregate {
    private long version = EventStorage.Constants.Versions.UNDEFINED_STREAM;

    @Override
    public boolean isDefined() {
        return version != EventStorage.Constants.Versions.UNDEFINED_STREAM;
    }

    @Override
    public void setVersion(Version version) {
        this.version = version.value();
    }

    @Override
    public Version version() {
        return new Version(version);
    }
}
