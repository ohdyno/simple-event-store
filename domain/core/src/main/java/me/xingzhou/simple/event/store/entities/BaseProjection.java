package me.xingzhou.simple.event.store.entities;

import java.time.Instant;
import me.xingzhou.simple.event.store.ids.RecordId;
import me.xingzhou.simple.event.store.storage.EventStorage;

public abstract class BaseProjection implements Projection {
    private long lastRecordId = EventStorage.Constants.Ids.UNDEFINED;
    private Instant lastUpdatedOn = EventStorage.Constants.InsertedOnTimestamps.NEVER;

    @Override
    public boolean isDefined() {
        return lastUpdatedOn != EventStorage.Constants.InsertedOnTimestamps.NEVER;
    }

    @Override
    public RecordId lastRecordId() {
        return new RecordId(lastRecordId);
    }

    @Override
    public Instant lastUpdatedOn() {
        return lastUpdatedOn;
    }

    @Override
    public void setLastRecordId(RecordId id) {
        this.lastRecordId = id.id();
    }

    @Override
    public void setLastUpdatedOn(Instant instant) {
        this.lastUpdatedOn = instant;
    }
}
