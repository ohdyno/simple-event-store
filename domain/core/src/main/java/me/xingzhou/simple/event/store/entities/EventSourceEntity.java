package me.xingzhou.simple.event.store.entities;

public interface EventSourceEntity {
    String APPLY_METHOD_NAME = "apply";

    default boolean isDefined() {
        return false;
    }
}
