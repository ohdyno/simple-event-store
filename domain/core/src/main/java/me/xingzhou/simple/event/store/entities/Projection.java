package me.xingzhou.simple.event.store.entities;

import java.time.Instant;
import java.util.Collections;
import java.util.List;
import me.xingzhou.simple.event.store.ids.RecordId;
import me.xingzhou.simple.event.store.ids.StreamName;

public interface Projection extends EventSourceEntity {
    RecordId lastRecordId();

    Instant lastUpdatedOn();

    void setLastRecordId(RecordId id);

    void setLastUpdatedOn(Instant instant);

    default List<StreamName> streamNames() {
        return Collections.emptyList();
    }
}
