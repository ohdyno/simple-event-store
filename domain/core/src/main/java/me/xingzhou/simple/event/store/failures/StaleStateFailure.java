package me.xingzhou.simple.event.store.failures;

public class StaleStateFailure extends RuntimeException {}
