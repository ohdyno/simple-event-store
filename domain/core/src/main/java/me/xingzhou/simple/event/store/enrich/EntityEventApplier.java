package me.xingzhou.simple.event.store.enrich;

import static me.xingzhou.simple.event.store.internal.tooling.CheckedExceptionHandlers.handleExceptions;

import me.xingzhou.simple.event.store.entities.EventSourceEntity;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class EntityEventApplier {

    private static final Logger log = LoggerFactory.getLogger(EntityEventApplier.class);
    private final EventTypesExtractor eventTypesExtractor;

    public EntityEventApplier(EventTypesExtractor eventTypesExtractor) {
        this.eventTypesExtractor = eventTypesExtractor;
    }

    public <T extends EventSourceEntity> void apply(EventRecord record, T entity) {
        log.debug("apply {} to entity {}", record, entity);
        handleExceptions(() -> {
            var parameterType = getMethod(entity, record.event().getClass());
            try {
                log.debug("attempting to find method {}({})", EventSourceEntity.APPLY_METHOD_NAME, parameterType);
                var method = entity.getClass().getMethod(EventSourceEntity.APPLY_METHOD_NAME, parameterType);

                log.debug("attempting to invoke method {}({})", EventSourceEntity.APPLY_METHOD_NAME, parameterType);
                method.invoke(entity, record.event());
            } catch (NoSuchMethodException e) {
                log.debug("unable to find method {}({})", EventSourceEntity.APPLY_METHOD_NAME, parameterType);

                log.debug(
                        "attempting to find public method {}({}, {})",
                        EventSourceEntity.APPLY_METHOD_NAME,
                        parameterType,
                        record.details().getClass());
                var method = entity.getClass()
                        .getMethod(
                                EventSourceEntity.APPLY_METHOD_NAME,
                                parameterType,
                                record.details().getClass());

                log.debug(
                        "attempting to invoke method {}({}, {})",
                        EventSourceEntity.APPLY_METHOD_NAME,
                        parameterType,
                        record.details().getClass());
                method.invoke(entity, record.event(), record.details());
            }
        });
    }

    private Class<?> getMethod(EventSourceEntity entity, Class<?> event) throws NoSuchMethodException {
        log.debug(
                "attempting to find the first parameter of the first {}(<? extends {}>) method",
                EventSourceEntity.APPLY_METHOD_NAME,
                event);
        return eventTypesExtractor.extractTypes(entity).stream()
                .filter(klass -> klass.isAssignableFrom(event))
                .findFirst()
                .orElseThrow(() -> new NoSuchMethodException(
                        "No such method: public void " + EventSourceEntity.APPLY_METHOD_NAME + "(" + event.getName()
                                + ") on " + entity.getClass().getName()));
    }
}
