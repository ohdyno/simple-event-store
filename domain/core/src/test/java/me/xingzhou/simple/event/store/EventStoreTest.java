package me.xingzhou.simple.event.store;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

import com.github.valfirst.slf4jtest.LoggingEvent;
import com.github.valfirst.slf4jtest.TestLogger;
import com.github.valfirst.slf4jtest.TestLoggerFactory;
import java.time.Duration;
import java.util.stream.Collectors;
import me.xingzhou.simple.event.store.enrich.EntityEventApplier;
import me.xingzhou.simple.event.store.enrich.EventTypesExtractor;
import me.xingzhou.simple.event.store.entities.ProjectionRecorder;
import me.xingzhou.simple.event.store.entities.TestAggregate;
import me.xingzhou.simple.event.store.events.TestEvent;
import me.xingzhou.simple.event.store.events.TestEventTypeConverter;
import me.xingzhou.simple.event.store.failures.StaleStateFailure;
import me.xingzhou.simple.event.store.serializer.adapters.JacksonEventSerializer;
import me.xingzhou.simple.event.store.storage.EventStorage;
import me.xingzhou.simple.event.store.storage.adapters.InMemoryEventStorage;
import org.approvaltests.Approvals;
import org.approvaltests.core.Options;
import org.approvaltests.scrubbers.DateScrubber;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import reactor.test.StepVerifier;

class EventStoreTest {

    private static final TestLogger logger = TestLoggerFactory.getTestLogger(EventStore.class);

    private EventStore store;

    @Test
    void enrichAProjection() {
        var event = new TestEvent("event-id");
        store.save(event, new TestAggregate());

        var recorder = store.enrich(new ProjectionRecorder());

        assertThat(recorder.isDefined()).isTrue();
        assertThat(recorder.appliedEvents()).hasSize(1);
        assertThat(recorder.appliedEvents().getFirst()).isEqualTo(event);
        assertThat(recorder.appliedEvents().getFirst()).isNotSameAs(event);

        StepVerifier.create(store.publisher())
                .assertNext(record -> {
                    assertThat(recorder.lastRecordId().id()).isEqualTo(record.id());
                    assertThat(recorder.lastUpdatedOn()).isEqualTo(record.insertedOn());
                })
                .thenCancel()
                .verify();
    }

    @Test
    void enrichAnAggregateFromAStream() {
        var event = new TestEvent("event-id");
        store.save(event, new TestAggregate());

        var recorder = store.enrich(new TestAggregate());

        assertThat(recorder.appliedEvents()).hasSize(1);
        assertThat(recorder.appliedEvents().getFirst()).isEqualTo(event);
        assertThat(recorder.appliedEvents().getFirst()).isNotSameAs(event);
        StepVerifier.create(store.publisher())
                .assertNext(record -> assertThat(recorder.version().value()).isEqualTo(record.version()))
                .thenCancel()
                .verify();
    }

    @Test
    void logging() {
        var event = new TestEvent("event-id");
        var aggregate = new TestAggregate();

        store.save(event, aggregate);
        store.save(event, aggregate);

        store.enrich(new ProjectionRecorder());

        var loggedEvents = logger.getAllLoggingEvents().stream()
                .map(LoggingEvent::toString)
                .collect(Collectors.joining("\n"));
        Approvals.verify(
                loggedEvents, new Options().withScrubber(DateScrubber.getScrubberFor("2025-04-20T04:21:48.191772Z")));
    }

    @Test
    void saveAnEvent() {
        var event = new TestEvent();
        var aggregate = store.save(event, new TestAggregate());

        assertThat(aggregate.isDefined()).isTrue();
        assertThat(aggregate.version().value()).isEqualTo(EventStorage.Constants.Versions.NEW_STREAM);
        assertThat(aggregate.appliedEvents()).containsExactly(event);
    }

    @Test
    void saveAnEventWithAnAggregateWithStaleVersion() {
        var event = new TestEvent();
        store.save(event, new TestAggregate());

        assertThatThrownBy(() -> store.save(event, new TestAggregate())).isInstanceOf(StaleStateFailure.class);
    }

    @BeforeEach
    void setUp() {
        logger.clearAll();
        var converter = new TestEventTypeConverter();
        var extractor = new EventTypesExtractor(converter);
        var serializer = new JacksonEventSerializer(converter);
        var storage = new InMemoryEventStorage();
        var applier = new EntityEventApplier(extractor);
        this.store = EventStore.build(new EventStoreDependencies(storage, serializer, extractor, applier));
        StepVerifier.setDefaultTimeout(Duration.ofMillis(100));
    }

    @Test
    void subscribe() {
        var event = new TestEvent();

        store.save(event, new TestAggregate());

        StepVerifier.create(store.publisher())
                .assertNext(record -> assertThat(record.id()).isEqualTo(EventStorage.Constants.Ids.START))
                .thenCancel()
                .verify();
    }
}
