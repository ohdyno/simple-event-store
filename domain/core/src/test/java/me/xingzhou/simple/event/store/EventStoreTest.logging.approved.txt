LoggingEvent{level=DEBUG, mdc={}, markers=[], keyValuePairs=[], throwable=Optional.empty, message='save event {} for aggregate {}', arguments=[TestEvent[id=event-id], TestAggregate{id='test-aggregate-id', appliedEvents=[TestEvent[id=event-id], TestEvent[id=event-id]]}]}
LoggingEvent{level=DEBUG, mdc={}, markers=[], keyValuePairs=[], throwable=Optional.empty, message='publish event record: {}', arguments=[StoredRecord[id=1, streamName=test-aggregate-id, eventType=TestEvent, eventContent={"id":"event-id"}, version=0, insertedOn=[Date1]]]}
LoggingEvent{level=DEBUG, mdc={}, markers=[], keyValuePairs=[], throwable=Optional.empty, message='save event {} for aggregate {}', arguments=[TestEvent[id=event-id], TestAggregate{id='test-aggregate-id', appliedEvents=[TestEvent[id=event-id], TestEvent[id=event-id]]}]}
LoggingEvent{level=DEBUG, mdc={}, markers=[], keyValuePairs=[], throwable=Optional.empty, message='publish event record: {}', arguments=[StoredRecord[id=2, streamName=test-aggregate-id, eventType=TestEvent, eventContent={"id":"event-id"}, version=1, insertedOn=[Date2]]]}
LoggingEvent{level=INFO, mdc={}, markers=[], keyValuePairs=[], throwable=Optional.empty, message='enriching {}', arguments=[ProjectionRecorder{appliedEvents=[TestEvent[id=event-id], TestEvent[id=event-id]]}]}
LoggingEvent{level=INFO, mdc={}, markers=[], keyValuePairs=[], throwable=Optional.empty, message='retrieved {} events', arguments=[2]}
LoggingEvent{level=INFO, mdc={}, markers=[], keyValuePairs=[], throwable=Optional.empty, message='applied {} events to {}', arguments=[2, ProjectionRecorder{appliedEvents=[TestEvent[id=event-id], TestEvent[id=event-id]]}]}
